src/mds/SimpleLock.h

```diff
@@ -325,7 +325,8 @@ class SimpleLock {
   bool can_read(client_t client) const {
     return get_sm()->states[state].can_read == ANY ||
       (get_sm()->states[state].can_read == AUTH && parent->is_auth()) ||
-      (get_sm()->states[state].can_read == XCL && client >= 0 && get_xlock_by_client() == client);
+      (get_sm()->states[state].can_read == XCL && client >= 0 && get_xlock_by_client() == client) ||
+      (get_sm()->states[state].can_read == XCL && client >= 0 && (get_xlock_by_client() == -1 && is_rdlocked()));
```

Comment:
would this logic also apply to `can_read_projected()`, which is right blow `can_read()` ?

Reply:
you are right, this logic also suitable for can_read_projected().

Reply:
Do you have the logs ? Are you sure the `filelock` state was in the `LOCK_XLOCKDONE` when this happened ?

As you mentioned in the tracker:

`t6. write op of setattr journal come back and triggered safe_reply of setattr. Notice： the filelock is XLOCK_DONE still but it’s xlocker is -1.`

after this, the `xlock` will be dropped by the setattr mdr and at the same time will drop the xlock_client, then in theory the `filelock` should be switched to another state.

What has suspend this to do the state transition ?

Reply:
> Do you have the logs ? Are you sure the `filelock` state was in the `LOCK_XLOCKDONE` when this happened ?
> 
> As you mentioned in the tracker:
> 
> `t6. write op of setattr journal come back and triggered safe_reply of setattr. Notice： the filelock is XLOCK_DONE still but it’s xlocker is -1.`
> 
> after this, the `xlock` will be dropped by the setattr mdr and at the same time will drop the xlock_client, then in theory the `filelock` should be switched to another state.
> 
> What has suspend this to do the state transition ?

I have updated logs in tracker 54421. 
The filelock lock sate can’t be transferred if its rd/wr/x_lock has been held.

Reply:
I am afraid you are changing the semantic of the `XCL`:

```
#define XCL  3 // auth or exclusive client
```

From my understanding for the `XCL` it should be:

```
- (get_sm()->states[state].can_read == XCL && client >= 0 && get_xlock_by_client() == client);
+ (get_sm()->states[state].can_read == XCL && ((client >= 0 && get_xlock_by_client() == client) || parent->is_auth());
```

@vshankar @gregsfortytwo Any idea about this ?

Reply:
So the MDS can read out of these states if it has a read lock, but if the exclusive client has a conflicting cap the MDS may not have authoritative state even if it's the auth MDS until after revoking the client's caps.

I don't remember if an xlocker client can buffer writes on SimpleLock or not, but that's the condition I worry about when changing this code.

Reply:
> So the MDS can read out of these states if it has a read lock, but if the exclusive client has a conflicting cap the MDS may not have authoritative state even if it's the auth MDS until after revoking the client's caps.
> 
> I don't remember if an xlocker client can buffer writes on SimpleLock or not, but that's the condition I worry about when changing this code.

We relax the check condition Only if the get_xlock_by_client() == -1. If don’t have xlocker on this inode,  can we ignore whether xlocker will buffer writes on SimpleLock?